### ISSUE TITLE
Block parser to drop events issue

### ISSUE THREAD

**Issue by RyanKershawWhittle** [splunk/splunk-connect-for-syslog#2162]
Hi,

I am having issues implementing a sc4s block parser to drop events:

block parser name_name_vmware-postfilter() {
   channel {
       #In this case the outcome is drop the event other logic such as adding indexed fields or editing the message is possible
       rewrite(r_set_dest_splunk_null_queue);
  };
};
application name_name_vmware-postfilter[sc4s-postfilter] {
filter {
       "${fields.sc4s_vendor}" eq "vmware" and
       "${fields.sc4s_product}" eq "vsphere"
       #Note regex will capture and drop events based on the keywords below.
       and message('^- - (lookupsvc-localhost_access|vpxd-main|statsmonitor|vmdird|sps|trustmanagement-svcs|vpxd-svcs-perf|sps-gc|osfsd|sso-tomcat|wcpsvc|vmafdd|vpxd-svcs-access|cis-license|ui-threadmonitor|ssoadminserver|analytics|content-library|vapi-endpoint-access|lookupsvc-gc|vmcad|sca-vmon\.std|ui-main|vapi-endpoint|trustmanagement-gc|vmon|ui-access|sca-gc|applmgmt-audit|observability-main|vapi-runtime|vapi-gc|vpxd-profiler|ui-gc|ui-runtime|ui-apigw|vapi-vcentershim|vum-vmacore|envoy-access) - -');
   };
   parser { name_name_vmware-postfilter(); };
};

The regex looks to capture keywords at the start of each event. Regex has been validated and works fine. Its worth mentioning I have tried several different iterations of regex to get this working and none seem to work with the parser..

If i comment # out the 'and message' line it works fine and drop events based on vendor and product.

please can you provide guidance on this.

**Comment by mstopa-splunk** [comment]
Hi @RyanKershawWhittle, it may be caused by nilvalues `-` which may not be part of the message macro. Please try regex with literals `(lookupsvc-localhost_access|vpxd-main...` but without nilvalues `-` and without matching beginning (or end) of the string. If this doesn't unblock you please send one or two raw log message examples for me to reproduce and check.

**Comment by RyanKershawWhittle** [comment]
Hi @mstopa-splunk 

removed the - as advised.. i.e. (lookupsvc.localhost_access|vpxd.main|statsmonitor|vmdird|sps|trustmanagement.svcs|vpxd.svcs.perf|sps.gc|osfsd|sso.tomcat|wcpsvc|vmafdd|vpxd.svcs.access|cis.license|ui.threadmonitor|ssoadminserver|analytics|content.library|vapi.endpoint.access|lookupsvc.gc|vmcad|sca.vmon\.std|ui.main|vapi.endpoint|trustmanagement.gc|vmon|ui.access|sca.gc|applmgmt.audit|observability.main|vapi.runtime|vapi.gc|vpxd.profiler|ui.gc|ui.runtime|ui.apigw|vapi.vcentershim|vum.vmacore|envoy.access)

unfortunately still no luck.

also tried a different approach:

block parser piaas_vxrail_vmware-postfilter() {

    channel {

        #In this case the outcome is drop the event other logic such as adding indexed fields or editing the message is possible

        rewrite(r_set_dest_splunk_null_queue);

   };

};

application piaas_vxrail_vmware-postfilter[sc4s-postfilter] {

filter {

        "${fields.sc4s_vendor}" eq "vmware" and

        "${fields.sc4s_product}" eq "vsphere"

        and not message ('^(?:(?!vpxd.\d).)*$');

    };

    parser { piaas_vxrail_vmware-postfilter(); };

};

still no luck.

Example logs:

- - vpxd-main - - -2023-09-07T09:37:12.886+01:00 info vpxd[3433757] [Originator@6876 sub=vpxLro opID=0b6cf675-d9] [VpxLRO] -- FINISH lro-462638005

- - vpxd 710336 - - Event [11156356] [1-1] [2023-09-07T08:37:29.086529Z] [name] [error] [name] [] [11156356] [Cannot login name]
-
- - sps - - -2023-09-07T09:37:48.621+01:00 [pool-3-thread-2] INFO  opI334d= name - [VLSI] Request took 0 millis to execute.

directory of parser - /opt/sc4s/local/config/app_parsers/rewriters

The idea is to drop all logs BUT 'vpxd digit'. 'digit' representing numerous digits after vpxd (i.e. vpxd 1453, vpxd 56724).

**Comment by mstopa-splunk** [comment]
The reason is that `vpxd` is not part of the `message`, it goes to `program`.

I checked the following to correctly filter out vmware vsphere events different than vpxd:
```
block parser name_name_vmware-postfilter() {
    channel {
        rewrite(r_set_dest_splunk_null_queue);
    };
};

application name_name_vmware-postfilter[sc4s-postfilter] {
    filter {
        "${fields.sc4s_vendor}" eq "vmware" 
        and "${fields.sc4s_product}" eq "vsphere"
        and not program('vpxd');
    };
    parser { name_name_vmware-postfilter(); };
};
```

You can find programs here: https://github.com/splunk/splunk-connect-for-syslog/blob/main/package/etc/conf.d/conflib/syslog/app-syslog-vmware_vsphere.conf

Please let me know if that solves the issue or process id `(i.e. vpxd 1453, vpxd 56724)` should be also included to the filter.

**Comment by RyanKershawWhittle** [comment]
Hi @mstopa-splunk

Progress! So it is now dropping everything that isnt 'vpxd'

However, its including vpxd-main, vpxd-svcs-access , and vpxd-svcs-perf. I wish for these to be dropped also. We just want to keep 'vpxd' 

I did try:
 and program ('vpxd-main')
 and not program ('vpxd');
 
 but it didnt work.

![image](https://github.com/splunk/splunk-connect-for-syslog/assets/101280607/9ecb9a4b-b1f1-4da6-bb0c-bc2c92305b71)

example logs:


- - vpxd-main - - -2023-09-07T16:23:22.998+01:00 info vpxd[19905] [Originator@6876 sub=vpxLro opID=00a800fa-c702-426a-94e8-6cdcbff6fb43-HMSINT-6143263-c0] [Vss24fpxdLfRO] -- FINISH lro-463292765
- 
- - vpxd-svcs-access - - -- 2023-09-07T16:29:37.553+0100 47982 431 tomcat-exec-202 200 "VMware vim-java 1.0" POST /invsvc/vmomi/sdk/ HTTP/1.1 1
- 
 - - vpxd-svcs-perf - - -2023-09-07T16:30:14.902+01:00 [tomcat-exec-60 INFO com.vmware.cis.serdd3ver.util.PerfLog opId=] Operation hasUserPrivilegesBatch for (user: blank\blank-1843d1d34cd-db7b-4e8c-a436-ceb9ddd93f6a6ea, request size: 1): took 0 ms.

**Comment by mstopa-splunk** [comment]
Hi @RyanKershawWhittle , I see. Let's turn off regex in the `program` filter by switching to string glob, no wildcards will result in the exact match:

```
block parser name_name_vmware-postfilter() {
    channel {
        rewrite(r_set_dest_splunk_null_queue);
    };
};

application name_name_vmware-postfilter[sc4s-postfilter] {
    filter {
        "${fields.sc4s_vendor}" eq "vmware" 
        and "${fields.sc4s_product}" eq "vsphere"
        and not program('vpxd' type("string") flags("global"));
    };
    parser { name_name_vmware-postfilter(); };
};
```

now it blocks `vpxd-main` but not `vpxd`

**Comment by RyanKershawWhittle** [comment]
Hi @mstopa-splunk 

This works!

**Comment by RyanKershawWhittle** [comment]
@mstopa-splunk 

If i want to include more programs to NOT be dropped what is the correct syntax:?

        and not program('vpxd' type("string") flags("global"));
        
        I.e. 'vpxd, vpxa'
              'vpxd vpxa'
              
           ??

**Comment by mstopa-splunk** [comment]
@RyanKershawWhittle `and not program` will work. The verbose way is good and self documenting but if you have more of them and they make regular patterns you can use wildcards. When you remove `flags("global")` you can use regex, which will be less efficient, but it's still successfully used in other filters

**Comment by RyanKershawWhittle** [comment]
Hi @mstopa-splunk 

Apologies let me be more clear. I want to keep 'hostd' and 'vpxa' programs. I have tried this config but it wont work:

```
block parser piaas_vmware_vcenter-postfilter() {

    channel {

        rewrite(r_set_dest_splunk_null_queue);

    };

};

 


 

application piaas_vmware_vcenter-postfilter[sc4s-postfilter] {

    filter {

        "${fields.sc4s_vendor}" eq "vmware"

        and "${fields.sc4s_product}" eq "vsphere"

        and not program('vpxd' type("string") flags("global"));

        and not program('hostd' type("string") flags("global"));

        and not program('vpxa' type("string") flags("global"));

    };

    parser { piaas_vmware_vcenter-postfilter(); };

};

```





https://github.com/splunk/splunk-connect-for-syslog/blob/main/package/etc/conf.d/conflib/syslog/app-syslog-vmware_vsphere.conf

**Comment by mstopa-splunk** [comment]
Right, please try removing semicolons prior to `and`. if it doesn't help, please provide a raw example for either of those two for us to investigate

**Comment by RyanKershawWhittle** [comment]
Hi @mstopa-splunk 

Remove the ; didnt work. 

```
block parser piaas_vmware_vcenter-postfilter() {

    channel {

        rewrite(r_set_dest_splunk_null_queue);

    };

};

 


 

application piaas_vmware_vcenter-postfilter[sc4s-postfilter] {

    filter {

        "${fields.sc4s_vendor}" eq "vmware"

        and "${fields.sc4s_product}" eq "vsphere"

        and not program('vpxd' type("string") flags("global"));

        and not program('hostd' type("string") flags("global"))

        and not program('vpxa' type("string") flags("global"))

    };

    parser { piaas_vmware_vcenter-postfilter(); };

};
```



example logs:



- - Vpxa - - [Originator@6876 sub="vpxLro" opID="HB-host-31@8946840-6a6e49fd-ec"][VpxLRO] -- FINISH lro-7598181


- - Hostd - - [Originator@6876 sub="Libs" opID="b2c6a5c4"]memory per disk MB: 81

**Comment by RyanKershawWhittle** [comment]
@mstopa-splunk thank you very much that worked!

Will close this. 

Thanks once again
